File: flink-cep/src/main/scala/org/apache/flink/dynamic/json/CepConverter.scala
package org.apache.flink.dynamic.json

import org.apache.flink.cep.pattern.{Pattern, Quantifier}
import org.apache.flink.dynamic.impl.json.serde.{ConditionSpecStdDeserializer, NodeSpecStdDeserializer, PatternTimesStdDeserializer, PatternTimesStdSerializer}
import org.apache.flink.dynamic.impl.json.spec.{ConditionSpec, GraphSpec, NodeSpec}
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.module.SimpleModule
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.datatype.jsr310.deser.DurationDeserializer
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.datatype.jsr310.ser.DurationSerializer

import java.time.Duration

/**
 * CepConverter
 *
 * @author chaoxin.lu
 * @version V 1.0
 * @since 2024-06-20
 */
object CepConverter {
  private val MAPPER: ObjectMapper = new ObjectMapper()
    .registerModule(new SimpleModule()
      .addDeserializer(classOf[ConditionSpec], ConditionSpecStdDeserializer.INSTANCE)
      .addDeserializer(classOf[Duration], DurationDeserializer.INSTANCE)
      .addDeserializer(classOf[NodeSpec], NodeSpecStdDeserializer.INSTANCE)
      .addDeserializer(classOf[Quantifier.Times], PatternTimesStdDeserializer.INSTANCE)
      .addSerializer(DurationSerializer.INSTANCE)
      .addSerializer(PatternTimesStdSerializer.INSTANCE))

  /**
   * toJSONString
   *
   * @param pattern
   * @return
   */
  def toJSONString[T, F <: T](pattern: Pattern[T, F]): String = {
    val graphSpec = GraphSpec.fromPattern(pattern)
    MAPPER.writeValueAsString(graphSpec)
  }

  /**
   * toPattern
   *
   * @param jsonString
   * @tparam T
   * @tparam F
   * @return
   */
  def toPattern[T, F <: T](jsonString: String): Pattern[T, F] = toPattern(jsonString, Thread.currentThread.getContextClassLoader)

  /**
   * toPattern
   *
   * @param jsonString
   * @tparam T
   * @tparam F
   * @return
   */
  def toPattern[T, F <: T](jsonString: String, userCodeClassLoader: ClassLoader): Pattern[T, F] = {
    if (userCodeClassLoader == null) {
      return toPattern(jsonString)
    }
    try {
      val deserializedGraphSpec: GraphSpec = MAPPER.readValue(jsonString, classOf[GraphSpec])
      deserializedGraphSpec.toPattern(userCodeClassLoader)
    } catch {
      case e: Exception => throw new IllegalStateException(e)
    }
  }

  /**
   * toGraphSpec
   *
   * @param jsonString
   * @return
   */
  def toGraphSpec(jsonString: String): GraphSpec = MAPPER.readValue(jsonString, classOf[GraphSpec])

  /**
   * toJSONString
   *
   * @param graphSpec
   * @return
   */
  def toJSONString(graphSpec: GraphSpec): String = MAPPER.writeValueAsString(graphSpec)

}

File: flink-cep/src/main/java/org/apache/flink/dynamic/impl/json/util/CepConverter.java
package org.apache.flink.dynamic.impl.json.util;

import org.apache.flink.cep.pattern.Pattern;
import org.apache.flink.cep.pattern.Quantifier;
import org.apache.flink.dynamic.impl.json.serde.ConditionSpecStdDeserializer;
import org.apache.flink.dynamic.impl.json.serde.NodeSpecStdDeserializer;
import org.apache.flink.dynamic.impl.json.serde.PatternTimesStdDeserializer;
import org.apache.flink.dynamic.impl.json.serde.PatternTimesStdSerializer;
import org.apache.flink.dynamic.impl.json.spec.ConditionSpec;
import org.apache.flink.dynamic.impl.json.spec.GraphSpec;
import org.apache.flink.dynamic.impl.json.spec.NodeSpec;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonProcessingException;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.module.SimpleModule;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.datatype.jsr310.deser.DurationDeserializer;
import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.datatype.jsr310.ser.DurationSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;

/**
 * Utils for translating a Pattern to JSON string and vice versa.
 */
public class CepConverter {

    private static final Logger LOG = LoggerFactory.getLogger(CepConverter.class);
    private static final ObjectMapper MAPPER = new ObjectMapper()
            .registerModule(new SimpleModule()
                    .addDeserializer(ConditionSpec.class, ConditionSpecStdDeserializer.INSTANCE)
                    .addDeserializer(Duration.class, DurationDeserializer.INSTANCE)
                    .addSerializer(Duration.class, DurationSerializer.INSTANCE)
                    .addDeserializer(NodeSpec.class, NodeSpecStdDeserializer.INSTANCE)
                    .addDeserializer(Quantifier.Times.class, PatternTimesStdDeserializer.INSTANCE)
                    .addSerializer(Quantifier.Times.class, PatternTimesStdSerializer.INSTANCE)
            );

    public static String toJSONString(Pattern<?, ?> pattern) throws JsonProcessingException {
        GraphSpec graphSpec = GraphSpec.fromPattern(pattern);
        return MAPPER.writeValueAsString(graphSpec);
    }

    public static <T, F extends T> Pattern<T, F> toPattern(String jsonString) {
        return toPattern(jsonString, Thread.currentThread().getContextClassLoader());
    }

    public static <T, F extends T> Pattern<T, F> toPattern(
            String jsonString,
            ClassLoader userCodeClassLoader) {
        if (userCodeClassLoader == null) {
            LOG.warn(
                    "The given userCodeClassLoader is null. Will try to use ContextClassLoader of current thread.");
            return toPattern(jsonString);
        }
        try {
            GraphSpec deserializedGraphSpec = MAPPER.readValue(jsonString, GraphSpec.class);
            return deserializedGraphSpec.toPattern(userCodeClassLoader);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public static GraphSpec toGraphSpec(String jsonString) throws JsonProcessingException {
        return MAPPER.readValue(jsonString, GraphSpec.class);
    }

    public static String toJSONString(GraphSpec graphSpec) throws JsonProcessingException {
        return MAPPER.writeValueAsString(graphSpec);
    }
}

